src/Illuminate/Auth/Access/Gate.php

```diff
@@ -302,7 +305,8 @@ public function authorize($ability, $arguments = [])
      */
     protected function raw($ability, $arguments = [])
     {
-        if (! $user = $this->resolveUser()) {
+        if (! ($user = $this->resolveUser()) &&
+            ! $this->allowsGuests($ability, Arr::wrap($arguments))) {
```

Comment:
Why are we checking this here right off the bat? Why isn't it enough that each step of the way decides whether it wants to handle `null` users?